hazelcast/src/hazelcast/client/spi/impl/ClientInvocation.cpp

```diff
@@ -0,0 +1,398 @@
+                                                                      objectName(objectName),
+                                                                      clientInvocationFuture(
+                                                                              new ClientInvocationFuture(logger, *this,
+                                                                                                         callIdSequence,
+                                                                                                         clientContext.getClientExecutionService())) {
+                }
+
+                ClientInvocation::ClientInvocation(spi::ClientContext &clientContext,
+                                                   std::auto_ptr<protocol::ClientMessage> &clientMessage,
+                                                   const std::string &objectName,
+                                                   const boost::shared_ptr<connection::Connection> &connection)
+                        : logger(
+                        util::ILogger::getLogger()),
+                          lifecycleService(
+                                  clientContext.getLifecycleService()),
+                          clientClusterService(
+                                  clientContext.getClientClusterService()),
+                          invocationService(
+                                  clientContext.getInvocationService()),
+                          executionService(
+                                  clientContext.getClientExecutionService()),
+                          clientMessage(
+                                  clientMessage),
+                          callIdSequence(
+                                  clientContext.getCallIdSequence()),
+                          partitionId(
+                                  UNASSIGNED_PARTITION),
+                          startTimeMillis(
+                                  util::currentTimeMillis()),
+                          objectName(
+                                  objectName),
+                          connection(
+                                  connection),
+                          clientInvocationFuture(new ClientInvocationFuture(
+                                  logger,
+                                  *this,
+                                  callIdSequence,
+                                  clientContext.getClientExecutionService())) {
+                }
+
+                ClientInvocation::ClientInvocation(spi::ClientContext &clientContext,
+                                                   std::auto_ptr<protocol::ClientMessage> &clientMessage,
+                                                   const std::string &objectName,
+                                                   boost::shared_ptr<connection::Connection> &connection,
+                                                   const boost::shared_ptr<Address> &address) : logger(
```

Comment:
This constructor looks odd. It has both connection and address. Connection seems unnecessary here.